rfcs/2020-09-02-3684-metric-namespaces.md

```diff
@@ -0,0 +1,177 @@
+# RFC 3684 - 2020-09-02 - First-class Metric Namespaces
+
+This RFC proposes making `namespace` a first-class field on the internal
+`Metric` type to allow it to be set in sources, manipulated in transforms, and
+used by sinks.
+
+## Scope
+
+This RFC will cover:
+
+- Separating the `namespace` of metrics into a separate field on `Metric`
+
+## Motivation
+
+As we add `metric` sources like
+[`apache_metrics`](https://github.com/timberio/vector/blob/master/rfcs/2020-08-21-3092-apache-metrics)
+and
+[`postgresql_metrics`](https://github.com/timberio/vector/blob/master/rfcs/2020-08-27-3603-postgres-metrics.md),
+that set their own namespaces (defaulting to `apache` and `postgresql`), it is
+becoming more clear that we may want to maintain the `namespace` separate from
+the metric name to allow for:
+
+- simple manipulation in transforms (e.g. as a filter)
+- to be sent as a separate field for sinks that require it (e.g.
+  `aws_cloudwatch_metrics`)
+- to be formatted differently depending on sink convention (e.g. it looks like
+  [NewRelic prefers
+  `namespace.name`](https://docs.newrelic.com/docs/telemetry-data-platform/get-data/apis/report-metrics-metric-api)
+  for metrics).
+
+I believe the current implementation proposals for these metrics sources will
+simply prefix the name as the `prometheus` and `statsd` sinks do, but this will
+be difficult to use with the `aws_cloudwatch_metrics` which requires the
+`namespace` as a separate field for the AWS API calls.
+
+Additionally, I think separating it will allow it to be more useful in
+transforms (users could currently emulate this by prefix matches of the metric
+name).
+
+## Internal Proposal
+
+Add `namespace` to
+[`Metric`](https://github.com/timberio/vector/blob/75844bc0f67d24ad1b54bfa130d074810ad2aa50/src/event/metric.rs#L10-L17):
+
+```rust
+pub struct Metric {
+  pub name: String,
+  pub namespace: Option<String>, // added
```

Comment:
It seems to me a fairly common use case will be wanting the namespace to be the same as the component name?

Reply:
:+1: or at least related (e.g. `apache_metrics` -> `apache`). We've been defaulting them for non-generic (generic being ones like `prometheus`) sources.